README.md
## Distributed Computation for NodeJS

How can you build a responsive and robust nodejs server that does some heavy
computational lifting?  Some node libraries (like the awesome [node-bcrypt][])
do their own threading internally and combine that with an async API.  This
allows libraries to internally thread their calls and use multiple cores.

  [node-bcrypt]: https://github.com/ncb000gt/node.bcrypt.js

While this is pretty awesome, it is significant work for library implementors,
and as this pattern becomes rampant, the application author looses fine grained
control over the resource usage of their server as well as the relative priority
of compute tasks.

If you just naively run computation on the main evaluation thread, you're blocking
node.js from doing *anything else* and making your whole server unresponsive.

## The solution?

`node-compute-cluster` is a tiny abstraction around a group of
processes and the [built-in IPC][] introduced in NodeJS 0.6.x.  It provides a simple
API by which you can allocate and run work on a cluster of computation processes.
This allows you to perform multiprocessing at a more granular level, and produce
a responsive yet efficient computation server.

 [built-in IPC]: http://nodejs.org/docs/v0.6.3/api/all.html#child_process.fork

## Installation

    $ npm install compute-cluster

## Usage

First you write your main program:

    const computecluster = require('compute-cluster');
    
    // allocate a compute cluster
    var cc = new computecluster({
      module: './worker.js'
    });
    
    var toRun = 10
    
    // then you can perform work in parallel
    for (var i = 0; i < toRun; i++) {
      cc.enqueue({}, function(err, r) {
        if (err) console.log("an error occured:", err);
        else console.log("it's nice:", r);
        if (--toRun === 0) cc.exit();
      });
    };

Next you write your `worker.js` program:

    process.on('message', function(m) {
      for (var i = 0; i < 100000000; i++);
      process.send('complete');
    });

All done!  Now you're distributing your computational load across multiple processes.

## API

### Constructor - `new require('compute-cluster')(<options>);`

Allocates a computation cluster.  Options include:

  * `module` - **required** the path to the module to load
  * `max_processes` - the maximum number of processes to spawn (default is `ciel(#cpus * 1.25)`)
  * `max_backlog` - the maximum length of the backlog, -1 indicates no limit (default is 10 * max_processes)

Example:

    var cc = new require('compute-cluster')({
      module: './foo.js',
      max_backlog: -1
    });

### Event: 'error'

An error event will be emited in exceptional circumstances.  Like if a child crashes.
Catch error events like this:

    cc.on('error', function(e) { console.log('OMG!', e); });

Default behavior is to exit on error if you don't catch.

### Events: 'debug' or 'info'

Events raise that hold an english, developer readable string describing
the state of the implementation.

### cc.enqueue(<args>, [cb])

enqueue a job to be run on the next available compute process, spawning one
if required (and `max_processes isn't hit).

args will be passed into the process (available via `process.on('message', ...)`).

`cb` is optional, and will be invoked with two params, `err` and `response`.
`err` indicates hard errors, response indicates successful roundtrip to the
compute process and is whatever the decided to `process.send()` in response. 

### cc.exit([cb])

Kill all child processes, invoking callback (with err param) when complete.

## LICENSE

Copyright (c) 2011, Lloyd Hilaiel <lloyd@hilaiel.com>

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted, provided that the above
copyright notice and this permission notice appear in all copies.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.


ChangeLog
0.0.5:
	* fix bug - queued work should be started even if enqueue is invoked from a completion callback.

0.0.4:
	* explicitly relay env to children, as default of child_process.fork seems to be not to do this

0.0.3:
	* fix github url in package.json

0.0.2:
	* the callback to .exit() really is optional

0.0.1:
	* the beginning of time


test/basic-test.js
#!/usr/bin/env node

const
vows = require('vows'),
assert = require('assert'),
computeCluster = require('../lib/compute-cluster'),
path = require('path');

var suite = vows.describe('basic tests');

// disable vows (often flakey?) async error behavior
suite.options.error = false;

suite.addBatch({
  "allocation of a compute cluster": {
    topic: function() {
      return computeCluster({
        module: path.join(__dirname, 'workers', 'echo.js')
      });
    },
    "runs without issue": function (cc) {
      assert.isObject(cc);
    },
    "and invocation against this cluster": {
      topic: function(cc) {
        var cb = this.callback;
        cc.enqueue("hello", function(e, r) {
          cb.call(self, { cc: cc, r: r });
        });

      },
      "succeeds": function (r) {
        assert.equal(r.r, 'hello');
      },
      "finally, exit": {
        topic: function(r) {
          r.cc.exit(this.callback);
        },
        "also succeeds": function(err) {
          assert.isNull(err);
        }
      }
    }
  }
});

// run or export the suite.
if (process.argv[1] === __filename) suite.run();
else suite.export(module);


test/workers/echo.js
process.on('message', function(m) {
  process.send(m);
});


example/after.js
#!/usr/bin/env node

const
os = require('os'),
path = require('path'),
ComputeCluster = require('../lib/compute-cluster.js');

// allocate a compute cluster
var computeCluster = new ComputeCluster({
  module: path.join(__dirname, "after_worker.js")
});

// if you don't handle errors, they will take down the process
computeCluster.on('error', function(e) {
  process.stderr.write('unexpected error from compute cluster: ' + e + "\n");
  process.exit(1);
});

var workDone = 0;
var starttime = new Date();
var lastoutput = starttime;

var outstanding = 0;
const MAX_OUTSTANDING = os.cpus().length * 2;

function addWork() {
  while (outstanding < MAX_OUTSTANDING) {
    outstanding++;
    computeCluster.enqueue({}, function(err, r) {
      outstanding--;
      workDone++;

      if (lastoutput.getTime() + (3 * 1000) < (new Date()).getTime())
      {
        lastoutput = new Date();
        console.log((workDone / ((lastoutput - starttime) / 1000.0)).toFixed(2),
                    "units work performed per second");
      }

      addWork();
    });
  }
}

addWork();


lib/compute-cluster.js
const
util = require('util'),
path = require('path'),
child_process = require('child_process'),
events = require('events');

function ComputeCluster(options) {
  if (!options || typeof options.module !== 'string') {
    throw "missing required 'module' argument";
  }
  // is module a file?
  if (!path.existsSync(options.module)) {
    throw "module doesn't exist: " + options.module;
  }
  if (options.max_processes &&
      (typeof options.max_processes !== 'number' || options.max_processes < 1)) {
    throw "when provided, max_processes must be an integer greater than one";
  }
  if (options.max_backlog && typeof options.max_backlog != 'number') {
    throw "when provided, max_backlog must be a number";
  }

  events.EventEmitter.call(this);

  // an array of child processes
  this._kids = {};
  this._MAX_KIDS = (options.max_processes || Math.ceil(require('os').cpus().length * 1.25));
  this._work_q = [];
  this._exiting = false;
  this._exit_cb;
  this._module = options.module;
  // how long shall we allow our queue to get before we stop accepting work?
  // (negative implies no limit.  careful, there.)
  this._MAX_BACKLOG = options.max_backlog || this._MAX_KIDS * 10;
};

util.inherits(ComputeCluster, events.EventEmitter);

ComputeCluster.prototype._onWorkerExit = function(pid) {
  var self = this;
  return function (code) {
    // if _exiting is false, we don't expect to be shutting down!
    if (!self._exiting) {
      self.emit('error',
                "compute process (" + pid + ") dies with code: " + code);
    }

    self.emit('info', "compute process (" + pid + ") exits with code " + code);

    delete self._kids[pid];
    if (self._exit_cb) {
      if (Object.keys(self._kids).length === 0) {
        self._exit_cb(null);
        _exit_cb = undefined;
      }
    }
  }
};

ComputeCluster.prototype._getFreeWorker = function() {
  var self = this;

  for (var i in this._kids) {
    if (!this._kids[i].job) return this._kids[i];
  }

  // no workers!  can we spawn one?
  if (Object.keys(this._kids).length < this._MAX_KIDS) {
    var k = {
      worker: child_process.fork(
        this._module,
        [],
        { env: process.env }
      ) };
    k.worker.on('exit', this._onWorkerExit(k.worker.pid));
    this._kids[k.worker.pid] = k;

    this.emit('info', "spawned new worker process (" +  k.worker.pid + ") " +
              Object.keys(this._kids).length + "/" + this._MAX_KIDS + " processes running");

    return k;
  }
};

ComputeCluster.prototype._runWorkOnWorker = function(work, worker) {
  var self = this;
  this.emit("debug", "passing compute job to process " + worker.worker.pid);
  var startTime = new Date();
  worker.worker.once('message', function(m) {
    // clear the in-progress job
    var cb = worker.job.cb;
    delete worker.job;

    // start the next
    self._assignWork();

    // call our client's callback
    if (cb) cb(null, m);
    // emit some debug info
    self.emit("debug", "process " + worker.worker.pid + " completed work in " +
              ((new Date() - startTime) / 1000.0).toFixed(2) + "s");
  });
  worker.worker.send(work.job);
  worker.job = work;
};

  // assign as many work units from work_q as possible to avialable
  // compute processes
ComputeCluster.prototype._assignWork = function() {
  while (this._work_q.length > 0) {
    var worker = this._getFreeWorker();
    if (!worker) break;

    this._runWorkOnWorker(this._work_q.shift(), worker);
  }
};

ComputeCluster.prototype.enqueue = function(args, cb) {
  // backlog size check
  if (this._MAX_BACKLOG > 0 && this._work_q.length >= this._MAX_BACKLOG) {
    this.emit('info', "maximum work backlog hit (" + this._MAX_BACKLOG +
              ")!  cannot enqueue additional work!");
    var mb = this._MAX_BACKLOG;
    setTimeout(function() {
      cb("cannot enqueue work: maximum backlog exceeded (" + mb + ")");
    }, 0);
    return this;
  }

  this._work_q.push({ job: args, cb: cb });
  this._assignWork();
  return this;
};

ComputeCluster.prototype.exit = function(cb) {
  if (Object.keys(this._kids).length === 0) {
    if (cb) setTimeout(function() { cb(null); }, 0);
  } else {
    this._exiting = true;
    this._exit_cb = cb;
    this.emit('info', "exit called, shutting down " + Object.keys(this._kids).length +
              " child processes");
    for (var i in this._kids) {
      this._kids[i].worker.kill();
    }
  }
  return this;
};

module.exports = ComputeCluster;

